File: app/infrastructure/s3/awsclient/S3PostSigner.java
package infrastructure.s3.awsclient;

import java.io.UnsupportedEncodingException;
import java.util.Date;
import java.util.Map;

public interface S3PostSigner {
    String buildEndpoint(String bucketName);

    Map<String, String> presignForm(Date userSpecifiedExpirationDate, String bucketName, String key, String acl, Map<String, String> additionalMetadata)
            throws UnsupportedEncodingException;
}

File: app/infrastructure/s3/S3PostSigner.scala
package infrastructure.s3

import java.time.Instant

trait S3PostSigner {
  def buildEndpoint(bucketName: String): String

  def presignForm(uploadParameters: UploadParameters): Map[String, String]
}

case class UploadParameters(
  expirationDateTime: Instant,
  bucketName: String,
  key: String,
  acl: String,
  additionalMetadata: Map[String, String]
)

final case class AwsCredentials(
  accessKeyId: String,
  secretKey: String,
  sessionToken: Option[String]
)
